### PR TITLE
[Merged by Bors] - refactor: Improve lemmas about sets of intermediate size

### PR DESCRIPTION
**Pull request by YaelDillies**
The lemmas around here are a mess. Rename `Finset.exists_smaller_set` to `Finset.exists_subset_card_eq`, `Finset.exists_intermediate_set` to `Finset.exists_subsuperset_card_eq`, and add the missing third `Finset.exists_superset_card_eq`. Same for `Set`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Fintype/Card.lean [hunk 1]
```diff
@@ -781,6 +781,11 @@ noncomputable def Finset.equivOfCardEq {s : Finset α} {t : Finset β} (h : s.ca
 theorem Finset.card_eq_of_equiv {s : Finset α} {t : Finset β} (i : s ≃ t) : s.card = t.card :=
   (card_eq_of_equiv_fintype i).trans (Fintype.card_coe _)
 
+/-- We can shrink a set `s` to any smaller size. -/
+lemma FinseT.exists_superset_card_eq [Fintype α] {n : ℕ} {s : Finset α} (hsn : s.card ≤ n)
```
Comment (b-mehta) [inline Mathlib/Data/Fintype/Card.lean:785]:
```suggestion
lemma Finset.exists_superset_card_eq [Fintype α] {n : ℕ} {s : Finset α} (hsn : s.card ≤ n)
```

File: Mathlib/Combinatorics/SetFamily/Shadow.lean [hunk 2]
```diff
@@ -310,13 +310,8 @@ theorem mem_upShadow_iff_exists_mem_card_add :
...
   · rintro ⟨t, ht, hts, hcard⟩
-    obtain ⟨u, htu, hus, hu⟩ :=
-      Finset.exists_intermediate_set 1
-        (by
-          rw [add_comm, ← hcard]
-          exact add_le_add_left (succ_le_of_lt (zero_lt_succ _)) _)
-        hts
-    rw [add_comm] at hu
+    obtain ⟨u, htu, hus, hu⟩ := Finset.exists_subsuperset_card_eq hts 1 -(Nat.le_add_right ..)
+      (by rw [← hcard]; exact add_le_add_left (succ_le_of_lt (zero_lt_succ _)) _)
```
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/Shadow.lean:314]: Probably `omega` works here?

**Review by b-mehta — APPROVED** [review]
Haha, I think these lemmas were some of my earliest mathlib PRs. I was surprised at the time that the names were accepted, and I'm surprised they've lasted this long :)

**Comment by YaelDillies** [comment]
> I'm surprised they've lasted this long :)

I am a very powerful procrastiner 😇

**Review by b-mehta — APPROVED** [review]
bors d+

File: Mathlib/Data/Fintype/Card.lean [hunk 3]
```diff
@@ -781,6 +781,11 @@ noncomputable def Finset.equivOfCardEq {s : Finset α} {t : Finset β} (h : s.ca
 theorem Finset.card_eq_of_equiv {s : Finset α} {t : Finset β} (i : s ≃ t) : s.card = t.card :=
   (card_eq_of_equiv_fintype i).trans (Fintype.card_coe _)
 
+/-- We can shrink a set `s` to any smaller size. -/
```
Comment (b-mehta) [inline Mathlib/Data/Fintype/Card.lean:784]: This is growing the set, not shrinking it!

File: Mathlib/Data/Finset/Card.lean [hunk 4]
```diff
@@ -618,39 +618,42 @@ theorem filter_card_add_filter_neg_card_eq_card
   rw [← card_union_of_disjoint (disjoint_filter_filter_neg _ _ _), filter_union_filter_neg_eq]
 #align finset.filter_card_add_filter_neg_card_eq_card Finset.filter_card_add_filter_neg_card_eq_card
 
+/-- Given a subset `s` of a set `t`, of sizes at most and at least `n` respectively, there exists a
+set `u` of size `n` which is both a superset of `s` and a subset of `t`. -/
+lemma exists_subsuperset_card_eq (hst : s ⊆ t) (hsn : s.card ≤ n) (hnt : n ≤ t.card) :
+    ∃ u, s ⊆ u ∧ u ⊆ t ∧ card u = n := by
+  classical
+  obtain ⟨k, hk⟩ := Nat.le.dest hnt
```
Comment (b-mehta) [inline Mathlib/Data/Finset/Card.lean:626]: I guess there's a cleaner proof by induction on hnt,  but it's up to you which you pick.

Comment (YaelDillies) [reply Mathlib/Data/Finset/Card.lean:626]: Sadly I did not manage to get `Nat.le_induction` to work

**Comment by YaelDillies** [comment]
bors merge

**Comment by YaelDillies** [comment]
Uh, I'm not sure why it was canceled.

bors merge

File: Mathlib/Data/Finset/Card.lean (hunk 4 above)
Comment (b-mehta) [reply Mathlib/Data/Finset/Card.lean:626]: Here's the shorter proof I was hoping for:
```
  classical
  refine Nat.decreasingInduction' ?_ hnt ⟨t, by simp [hst]⟩
  intro k _ hnk ⟨u, hu₁, hu₂, hu₃⟩
  obtain ⟨a, ha⟩ : (u \ s).Nonempty := by rw [←card_pos, card_sdiff hu₁]; omega
  simp only [mem_sdiff] at ha
  exact ⟨u.erase a, by simp [subset_erase, erase_subset_iff_of_mem (hu₂ _), *]⟩
  ```

Comment (YaelDillies) [reply Mathlib/Data/Finset/Card.lean:626]: Ah, that's not induction *on* `hnt` but induction *using* `hnt`!

Comment (YaelDillies) [reply Mathlib/Data/Finset/Card.lean:626]: #14425

**Pull request closed without merging**